File: python/CardDeck.py
"""Representation of a standard 52-deck of French-suited playing cards.
"""

from random import randint
from CardSuit import SUITS
from CardValue import VALUES
from Card import Card


class CardDeck():
    """Representation of a standard 52-deck of playing cards.
    
    Attributes:

    """

    def __init__(self):
        """Initializes the deck of cards."""
        
        self.cards = []
        self.__newDeck__()
    

    def __newDeck__(self):
        suits = SUITS
        values = VALUES

        for suit in suits:
            for value in values:
                self.cards.append(Card(suit, value))
    

    def shuffle(self):
        """Shuffle the deck of cards in a realistic manner.

        This is done by first splitting the deck roughly in half, and then recombining
        the deck.
        """

        splitPoint = randint(20, 32)
        
        topHalf = self.cards[0:splitPoint]
        bottomHalf = self.cards[splitPoint:]

        self.__recombineDeck__(topHalf, bottomHalf)
    

    def __recombineDeck__(self, leftHalf, rightHalf):
        """
        Recombines the two halfs of the deck of cards in a realistic manner.

        This is done by randomly choosing between one of the two given halves of the
        deck and adding the card at the bottom of the chosen half to a "pile" of 
        cards. Then the process repeats by randomly choosing between the two halves 
        again.
        This process repeats until one of the two halves is depleted, at which point the
        other half is added to the top of the pile. This pile is then saved as the 
        freshly shuffled deck.

        Atrributes:
            leftHalf: list
                one of the two halves of the original deck
            rightHalf: list
                the other half of the original deck
        """

        newDeck = []

        while((len(leftHalf) > 0) and (len(rightHalf) > 0)):
            leftCard = (randint(0, 1) == 0)

            if(leftCard):
                newDeck.append(leftHalf.pop())
            else:
                newDeck.append(rightHalf.pop())
        
        if(len(leftHalf) > 0):
            newDeck = leftHalf + newDeck
        else:
            newDeck = rightHalf + newDeck
        
        self.cards = newDeck


    def reset(self):
        """Set the cards in the deck back to their original (unshuffled) order"""
        self.cards = sorted(self.cards)
    

    def __str__(self):
        deckMap = map((lambda card: str(card)), self.cards) 
        deckList = list(deckMap)
        return str(deckList) 
    

    def __iter__(self):
        self.__index = 0
        return self

    def __next__(self):
        try:
            currentCard = self.cards[self.__index]
            self.__index += 1
            return currentCard
        except IndexError:
            raise StopIteration

__author__ = "Garrett Kamila Crayton (garrett.crayton@gmail.com)"
__date__ = "2021 April 11"
__credits__ = "Garrett Kamila Crayton (garrett.crayton@gmail.com)"
__version__ = "2.0"

File: PreviousVersions/v3/src/CardDeck.java
import java.util.ArrayList;
import java.util.Random;

/**
 * Represents a standard deck of playing cards.
 *
 * @author Garrett Kamila Crayton
 * @version 1.0
 */
public class CardDeck
{
    /**
     * The "deck" of <code>Cards</code>
     */
    public ArrayList<Card> cards;
    private static Random random;

    /**
     * Construct a new CardDeck object.
     */
    public CardDeck()
    {
        init();
        newDeck();
    }

    /**
     * Initialize the class attributes (if necessary
     */
    private static void init()
    {
        if(random == null)
        {
            random = new Random();
        }
    }

    /**
     * Generate a new deck of cards.
     */
    private void newDeck()
    {
        cards = new ArrayList<Card>();
        for(Suit suit: Suit.values())
        {
            for(Rank rank: Rank.values())
            {
                cards.add(new Card(suit, rank));
            }
        }
    }

    /**
     * Shuffle the cards in the deck in a realistic manner.
     */
    public void shuffle()
    {
        // Split the deck
        ArrayList<ArrayList> halves = splitDeck();

        // Recombine the decks
        recombineDecks(halves.get(0), halves.get(1));
    }

    /**
     * Split the deck of cards roughly in half.
     *
     * @return  ArrayList containing both halves of the original deck
     */
    private ArrayList<ArrayList> splitDeck()
    {
        ArrayList<ArrayList> halves = new ArrayList<ArrayList>();
        ArrayList<Card> topHalf;
        ArrayList<Card> bottomHalf;

        int middle = (random.nextInt(12) + 20);
        topHalf = new ArrayList<Card>(cards.subList(0, middle));
        bottomHalf = new ArrayList<Card>(cards.subList(middle, cards.size()));

        halves.add(topHalf);
        halves.add(bottomHalf);

        return halves;
    }

    /**
     * Recombines both halves into one whole deck.
     *
     * @param leftHalf      one of the two "halves" of the original deck
     * @param rightHalf     the other "half" of the original deck
     */
    private void recombineDecks(ArrayList<Card> leftHalf, ArrayList<Card> rightHalf)
    {
        cards.clear();

        while((leftHalf.size() != 0) && (rightHalf.size() != 0))
        {
            boolean choseLeft = random.nextBoolean();

            if(choseLeft)
            {
                cards.add(0, leftHalf.remove(leftHalf.size() - 1));
            }
            else
            {
                cards.add(0, rightHalf.remove(rightHalf.size() - 1));
            }
        }

        if(leftHalf.isEmpty())
        {
            cards.addAll(0, rightHalf);
        }
        else
        {
            cards.addAll(0, leftHalf);
        }
    }

    /**
     * Reset the deck (sort cards into original order)
     */
    public void reset()
    {
        cards.sort(Card.comparator);
    }

    /**
     * String representation of the deck of cards
     *
     * @return  the string representation of the deck of cards
     */
    public String toString()
    {
        return cards.toString();
    }
}
